crates/project-model/src/project_json.rs

```diff
@@ -99,26 +98,24 @@ impl ProjectJson {
     /// * `data` - The parsed contents of `rust-project.json`, or project json that's passed via
     ///            configuration.
     pub fn new(base: &AbsPath, data: ProjectJsonData) -> ProjectJson {
+        let absolutize =
+            |p| AbsPathBuf::try_from(p).unwrap_or_else(|path| base.join(&path)).normalize();
         ProjectJson {
-            sysroot: data.sysroot.map(|it| base.join(it)),
-            sysroot_src: data.sysroot_src.map(|it| base.join(it)),
+            sysroot: data.sysroot.map(absolutize),
```

Comment:
just

```
data.sysroot.map(|it| base.join(it)),
```

should do the trick. That's the semantics of Rust's join: if the path is absolute, base is ignored. It _is_ a bit surprising, perhaps we can benefit from

```
impl AbsPath {
    pub fn absolutize(&self, path: impl AsRef<Path>) -> AbsPathBuf;
    pub fn join(&self, path: &RelPath) -> AbsPathBuf;
}
impl RelPath {
    pub fn join(&self, path: &RelPath) -> RelPathBuf;
}
```